File: sdk/go/internal/nodeclient/computer_use.go
package nodeclient

import (
	"bytes"
	"context"

	nodesandboxv1 "github.com/cofy-x/axern/sdk/go/gen/axern/node/sandbox/v1"
)

type ComputerUseStatus struct {
	Available    bool
	Display      string
	Backend      string
	Reason       string
	Dependencies []ComputerUseDependencyStatus
}

type ComputerUseDependencyStatus struct {
	Name      string
	Available bool
	Reason    string
}

type ComputerUseScreenshot struct {
	Data        []byte
	ContentType string
}

type ComputerUseRegion struct {
	X      int32
	Y      int32
	Width  int32
	Height int32
}

type ComputerUseScreenshotOptions struct {
	ShowCursor bool
	Region     *ComputerUseRegion
	Format     string
	Quality    int32
	Scale      float64
}

type ComputerUseDisplay struct {
	Display string
	Backend string
	Width   int32
	Height  int32
}

type ComputerUseMouseOptions struct {
	Action    string
	X         int32
	Y         int32
	ToX       int32
	ToY       int32
	Button    string
	Direction string
	Amount    int32
}

type ComputerUseKeyboardOptions struct {
	Text    string
	Key     string
	Keys    []string
	DelayMS int32
}

func (c *Client) ComputerUseStatus(ctx context.Context) (ComputerUseStatus, error) {
	response, err := c.nodes.ComputerUseStatus(ctx, &nodesandboxv1.ComputerUseStatusRequest{
		AllocationID: c.allocationID,
	})
	if err != nil {
		return ComputerUseStatus{}, err
	}
	return ComputerUseStatus{
		Available:    response.GetAvailable(),
		Display:      response.GetDisplay(),
		Backend:      response.GetBackend(),
		Reason:       response.GetReason(),
		Dependencies: computerUseDependencies(response.GetDependencies()),
	}, nil
}

func (c *Client) ComputerUseScreenshot(ctx context.Context, options ComputerUseScreenshotOptions) (ComputerUseScreenshot, error) {
	response, err := c.nodes.ComputerUseScreenshot(ctx, &nodesandboxv1.ComputerUseScreenshotRequest{
		AllocationID: c.allocationID,
		ShowCursor:   options.ShowCursor,
		Region:       computerUseRegion(options.Region),
		Format:       options.Format,
		Quality:      options.Quality,
		Scale:        options.Scale,
	})
	if err != nil {
		return ComputerUseScreenshot{}, err
	}
	return ComputerUseScreenshot{
		Data:        bytes.Clone(response.GetData()),
		ContentType: response.GetContentType(),
	}, nil
}

func (c *Client) ComputerUseDisplay(ctx context.Context) (ComputerUseDisplay, error) {
	response, err := c.nodes.ComputerUseDisplay(ctx, &nodesandboxv1.ComputerUseDisplayRequest{
		AllocationID: c.allocationID,
	})
	if err != nil {
		return ComputerUseDisplay{}, err
	}
	return ComputerUseDisplay{
		Display: response.GetDisplay(),
		Backend: response.GetBackend(),
		Width:   response.GetWidth(),
		Height:  response.GetHeight(),
	}, nil
}

func (c *Client) ComputerUseMouse(ctx context.Context, options ComputerUseMouseOptions) error {
	_, err := c.nodes.ComputerUseMouse(ctx, &nodesandboxv1.ComputerUseMouseRequest{
		AllocationID: c.allocationID,
		Action:       options.Action,
		X:            options.X,
		Y:            options.Y,
		ToX:          options.ToX,
		ToY:          options.ToY,
		Button:       options.Button,
		Direction:    options.Direction,
		Amount:       options.Amount,
	})
	return err
}

func (c *Client) ComputerUseKeyboard(ctx context.Context, options ComputerUseKeyboardOptions) error {
	_, err := c.nodes.ComputerUseKeyboard(ctx, &nodesandboxv1.ComputerUseKeyboardRequest{
		AllocationID: c.allocationID,
		Text:         options.Text,
		Key:          options.Key,
		Keys:         append([]string(nil), options.Keys...),
		DelayMs:      options.DelayMS,
	})
	return err
}

func computerUseRegion(region *ComputerUseRegion) *nodesandboxv1.ComputerUseRegion {
	if region == nil {
		return nil
	}
	return &nodesandboxv1.ComputerUseRegion{
		X:      region.X,
		Y:      region.Y,
		Width:  region.Width,
		Height: region.Height,
	}
}

func computerUseDependencies(items []*nodesandboxv1.ComputerUseDependencyStatus) []ComputerUseDependencyStatus {
	out := make([]ComputerUseDependencyStatus, 0, len(items))
	for _, item := range items {
		out = append(out, ComputerUseDependencyStatus{
			Name:      item.GetName(),
			Available: item.GetAvailable(),
			Reason:    item.GetReason(),
		})
	}
	return out
}

File: sdk/python/src/axern_sdk/sandbox/computer_use.py
"""Synchronous sandbox computer-use helpers."""

from __future__ import annotations

from typing import Protocol

from axern_sdk.node import (
    ComputerUseDisplay,
    ComputerUseRegion,
    ComputerUseScreenshot,
    ComputerUseStatus,
    NodeSandboxClient,
)


class _HasNodeClient(Protocol):
    def _node_client(self) -> NodeSandboxClient: ...


class SandboxComputerUseMixin:
    """Computer-use operations for a started sandbox."""

    def computer_use_status(self: _HasNodeClient, *, timeout_seconds: int = 30) -> ComputerUseStatus:
        return self._node_client().computer_use_status(rpc_timeout=timeout_seconds)

    def computer_use_screenshot(
        self: _HasNodeClient,
        *,
        show_cursor: bool = False,
        region: ComputerUseRegion | None = None,
        format: str = "",
        quality: int = 0,
        scale: float = 0,
        timeout_seconds: int = 30,
    ) -> ComputerUseScreenshot:
        return self._node_client().computer_use_screenshot(
            show_cursor=show_cursor,
            region=region,
            format=format,
            quality=quality,
            scale=scale,
            rpc_timeout=timeout_seconds,
        )

    def computer_use_display(self: _HasNodeClient, *, timeout_seconds: int = 30) -> ComputerUseDisplay:
        return self._node_client().computer_use_display(rpc_timeout=timeout_seconds)

    def computer_use_mouse(
        self: _HasNodeClient,
        *,
        action: str = "click",
        x: int = 0,
        y: int = 0,
        to_x: int = 0,
        to_y: int = 0,
        button: str = "",
        direction: str = "",
        amount: int = 0,
        timeout_seconds: int = 30,
    ) -> None:
        self._node_client().computer_use_mouse(
            action=action,
            x=x,
            y=y,
            to_x=to_x,
            to_y=to_y,
            button=button,
            direction=direction,
            amount=amount,
            rpc_timeout=timeout_seconds,
        )

    def computer_use_keyboard(
        self: _HasNodeClient,
        *,
        text: str = "",
        key: str = "",
        keys: list[str] | tuple[str, ...] = (),
        delay_ms: int = 0,
        timeout_seconds: int = 30,
    ) -> None:
        self._node_client().computer_use_keyboard(
            text=text,
            key=key,
            keys=keys,
            delay_ms=delay_ms,
            rpc_timeout=timeout_seconds,
        )
